File: BaekJoon Online Judge/C++/Step16-Queue and Deque/10866.cpp
//
//  10866.cpp
//  Step16-Queue and Deque
//
//  Created by 김예빈 on 2019. 9. 24..
//  Copyright © 2019년 김예빈. All rights reserved.
//

#include <iostream>
#include <string>
using namespace std;

int deque[10000];
int rear = 0;

void push_front(int);
void push_back(int);
int pop_front();
int pop_back();
bool empty();
int front();
int back();

int main(int argc, const char * argv[]) {
    cin.tie(NULL);
    ios::sync_with_stdio(false);
    
    int n, num;
    string ir;
    
    cin >> n;
    
    while(n--) {
        cin >> ir;
        
        if(ir == "push_front") {
            cin >> num;
            push_front(num);
        }
        
        if(ir == "push_back") {
            cin >> num;
            push_back(num);
        }
        
        if(ir == "pop_front")
            cout << pop_front() << "\n";
        
        if(ir == "pop_back")
            cout << pop_back() << "\n";
        
        if(ir == "size")
            cout << rear << "\n";
        
        if(ir == "empty")
            cout << empty() << "\n";
        
        if(ir == "front")
            cout << front() << "\n";
        
        if(ir == "back")
            cout << back() << "\n";
    }
    
    return 0;
}

void push_front(int input) {
    for(int i = rear; i > 0; i--)
        deque[i] = deque[i - 1];
    
    rear++;
    deque[0] = input;
}

void push_back(int input) {
    deque[rear++] = input;
}

int pop_front() {
    if(empty())
        return -1;
    else {
        int front = deque[0];
        rear--;
        
        for(int i = 0; i < rear; i++)
            deque[i] = deque[i + 1];
        
        return front;
    }
}

int pop_back() {
    if(empty())
        return -1;
    else {
        int back = deque[--rear];
        
        return back;
    }
}

bool empty() {
    if(rear == 0)
        return true;
    else
        return false;
}

int front() {
    if(empty())
        return -1;
    else
        return deque[0];
}

int back() {
    if(empty())
        return -1;
    else
        return deque[rear - 1];
}

File: BaekJoon Online Judge/Swift/Step18-Queues and Deques/10866.swift
//
//  10866.swift
//  Step18-Queues and Deques
//
//  Created by Yebin Kim on 2019/12/24.
//  Copyright © 2019 김예빈. All rights reserved.
//

import Foundation

// Use two stacks to implement deque
public struct Stack<T> {
    private var stack = Array<T>()
    
    public mutating func push(_ element: T) {
        stack.append(element)
    }
    
    public mutating func pop() -> T? {
        return stack.popLast()
    }
    
    public var size: Int {
        return stack.count
    }
    
    public var empty: Bool {
        return stack.isEmpty
    }
    
    public var top: T? {
        return stack.last
    }
    
    public mutating func reverse() -> Stack<T> {
        var ret = Stack<T>()
        var copy = stack
        
        for _ in 0..<copy.count {
            ret.push(copy.popLast()!)
        }
        
        return ret
    }
}

public struct Deque<T> {
    private var frontStack = Stack<T>()
    private var backStack = Stack<T>()
    
    public mutating func push_front(_ element: T) {
        frontStack.push(element)
        backStack = frontStack.reverse()
    }
    
    public mutating func push_back(_ element: T) {
        backStack.push(element)
        frontStack = backStack.reverse()
    }
    
    public mutating func pop_front() -> T {
        if frontStack.empty {
            return -1 as! T
        } else {
            let ret = frontStack.pop()
            backStack = frontStack.reverse()
            return ret!
        }
    }
    
    public mutating func pop_back() -> T {
        if backStack.empty {
            return -1 as! T
        } else {
            let ret = backStack.pop()
            frontStack = backStack.reverse()
            return ret!
        }
    }
    
    public var size: Int {
        return frontStack.size
    }
    
    public var empty: Int {
        if frontStack.empty {
            return 1
        } else {
            return 0
        }
    }
    
    public var front: T {
        if frontStack.empty {
            return -1 as! T
        } else {
            return frontStack.top!
        }
    }
    
    public var back: T {
        if backStack.empty {
            return -1 as! T
        } else {
            return backStack.top!
        }
    }
}

var deque = Deque<Int>()

for _ in 0..<Int(readLine()!)! {
    let op = readLine()!.split(separator: " ")
    
    switch op[0] {
    case "push_front":
        deque.push_front(Int(String(op[1]))!)
        break
        
    case "push_back":
        deque.push_back(Int(String(op[1]))!)
        break
        
    case "pop_front":
        print(deque.pop_front())
        break
        
    case "pop_back":
        print(deque.pop_back())
        break
        
    case "size":
        print(deque.size)
        break
        
    case "empty":
        print(deque.empty)
        break
        
    case "front":
        print(deque.front)
        break
        
    case "back":
        print(deque.back)
        break
        
    default: break
    }
}
